# Иванов Роман
# 98 Вариант
# Группа №9372
# Задача 2

# Постановка задачи

Зависимость объема выбросов вредных веществ $W(D)$ в окружающую среду (в год) от объемов
производства $D$ промышленного предприятия задается следующим соотношение

$W(D)$ = $max(0, 1.74e^(0.31D)−2.4)$

Способность окружающей среды к естественному рассеянию примесей зависит от объема
выбросов. Так, при небольших выбросах все вредные вещества могут быть рассеяны, а при
больших — часть вредных веществ остается до следующих периодов. Зависимость остаточных
загрязнений к следующему периоду (году) от общих загрязнений на данном периоде выража-
ется следующим соотношением: $A(P) = 0, при P <= 5$   $A(P) = 0.26e^(0.38P)−1.75, при$ $5 < P <= 10$   $A(P) = P, при P > 10$
    
    
Производство продукции приводит к получению дохода $B(D)$:
$B(D) = 6.17\log(D+1.4)−2.16$

Наличие в окружающей среде загрязнений приводит к дополнительным платежам C(P) (их
можно понимать либо как штрафы, либо как условные единицы комплексной целевой функции,
учитывающей как финансовую, так и социальную составляющие):
$C(P) = max(0, 0.08e^(0.33P)−0.15)$
Необходимо выбрать объемы производства на ближайшие $5$ лет, максимизирующие прибыль предприятия с учетом негативного влияния на экологию (при условии, что объем производства может выражаться только целым числом от $0$ до $6$, а остаточный уровень загрязнений
по истечении 5 лет не должен превышать $10$).

# 1 Шаг. Анализ задачи

В задаче задана периодизация процесса принятия решений: есть 5 точек принятия решений, которые соответствуют началам каждого из годов.
При анализе условия задачи необходимо выделить три составляющие, обязательные при решении задачи методом динамического программирования: выигрыш, управление и состояние.

Выигрыш в данной задаче соответствует доходу, получаемому от выбранного объема производства за вычетом расходов за загрязнения, оставшиеся в окружающей среде

Управление - выбранный объем производства

Состояние - состояние окружающей среды(остаточные загрязнения).

Задача состоит в том, чтобы максимизировать прибыль, не выйдя за пределы загрязнения

# 2 Шаг. Составление функции Белмана

В качестве этапов $i$ возьмем год производсва, где $i$ $\in$ $[0;4]$. 

Объем производства $d_i$, где $i$ $\in$ $[0;6]$.

Уровень выбросов будем обозначать $w_i' = W'(d_i)$.

Так как есть функция остаточных загрязнений к следующему году, то уровень выбросов будет считать следующим выражением

$a$$_i$$_+$$_1 = (A(a_i+w_i′))$. В самый первый год $a_0 = 0$, а в последний 5 год $a_4 \leqslant 10$

Доход в 𝑖-ый год будет равен $b_i = 𝐵(d_i)$, а штраф $ci = C(a_i + w_i′)$, тогда
итоговая прибыль за год равна $(b_i − c_i)$.


**Перейдем к самому составлению уравнения Белмана**

Состояниями $S_i$ будут выбросы за года с 0 до stage включительно: $a_i$, 

управлениями $u_i$ – объемы производства: $d_i$, 

выигрышем $w_i$ – доход в i-ый год: 

\begin{equation}w_i = (b_i − c_i) = B(u_i) − C(S_i + w_i') = B(u_i) − C(S_i + W'(u_i))\end{equation}. 

Функция изменения состояния  $𝜑(S_i, u_i)$  – уровень выбросов для следующего 

\begin{equation} 𝜑(S_i, u_i) = a_{i_1} =A(S_i + w_i) = A(S_i + W^′(u_i))\end{equation}. 

Тогда уравнение Белмана примет следующий вид:

\begin{equation} W_i(S) = \max\limits_{u_i\in{ \{u|u\in[0;6], A(S_i+W'(u_i))\leqslant10} \}} {\{ B(u_i) − C(S_i + W'(u_i)) + W_{i+1}(A(S_i + W'(u_i))) }\} \end{equation}

# Шаг 3. Решение задачи

Для удобства напишем отделеные функции для подсчета выбросов, остаточных загрязнений, прибыли и расходов.

In [2]:
import math

#функция для подсчета выбросов в зависимости от выбранного объема производства
def W(D):
    return max(0, 1.74 * (math.e ** (0.31 * D)) - 2.4)

#функция для подсчета остаточных хагрязнений
def A(P):
    if P <= 5:
        return 0
    elif P <= 10:
        return 0.26 * (math.e ** (0.38 * P)) - 1.75
    else:
        return P
    
#функция для подсчета прибыли от выбранного объема производства
def B(D):
    return 6.17 * math.log(D + 1.4) - 2.16

#функция для подсчета штрафа за загрязнения
def C(P):
    return max(0, 0.08 * (math.e ** (0.33 * P)) - 0.15)

In [3]:
from typing import Callable

class OptimalVolumeSolver:
    def __init__(self, W : Callable[[int], float], #функция для подсчета выбросов в зависимости от выбранного объема производства
                       A : Callable[[float], float], #функция для подсчета остаточных хагрязнений
                       B : Callable[[int], float], #функция для подсчета прибыли от выбранного объема производства
                       C : Callable[[float], float], #функция для подсчета штрафа за загрязнения
                       capacity : int, # макисмальный остаточный уровень загрязнений к концу этапов производства
                       stagesCount : int, #количество этапов производства (лет)
                       controlsRange : range): #диапазон объемов производства
        self.w = W
        self.b = B
        self.c = C
        self.a = A
        self.capacity = capacity
        self.stagesCount = stagesCount
        self.controlsRange = controlsRange
        self.W_cache = [{} for _ in range(stagesCount)]

In [14]:
class OptimalVolumeSolver(OptimalVolumeSolver):
    #Расчет функции Беллмана
    def W(self, stage : int, state : float):
        '''
        stage - этап - год производства [0; stagesCount)
        state - состояние - количество 'накопленных выбросов' в предыдущих этапах [0; stage - 1]
        u - управление - объем производства - controlsRange
        '''
                
        if stage >= self.stagesCount:
            return (0, None)
        
        if state in self.W_cache[stage]:
            return self.W_cache[stage][state]

        best_w = None
        best_u = None
        for u in self.controlsRange:
            stage_new_state = state + self.w(u)
            next_stage_state = self.a(stage_new_state)
            if next_stage_state <= self.capacity:  
                wi = self.b(u) - self.c(stage_new_state) + self.W(stage + 1, next_stage_state)[0]
                if best_w is None or wi > best_w:
                    best_w = wi
                    best_u = u

        self.W_cache[stage][state] = (best_w, best_u)
        return best_w, best_u

In [15]:
class OptimalVolumeSolver(OptimalVolumeSolver):
    def restore_optimal(self):
        control = []
        state = 0
        for stage in range(self.stagesCount):
            u = self.W(stage, state)[1]
            control.append(u)
            state = self.a(state + self.w(u))
        return control

    def solve(self):
        return self.W(0, 0)[0], self.restore_optimal()

In [21]:
result = OptimalVolumeSolver(W, A, B, C, 10, 5, range(0, 6))
result.solve()

(43.368080413179236, [5, 5, 4, 5, 5])

Получили оптимальную прибыль в **43.368** единиц, при этом объем производства для каждого года следующий

|Год    |Объем производства|
|-------|------------------|
|**1**  |5                 |
|**2**  |5                 |
|**3**  |4                 |                 
|**4**  |5                 |
|**5**  |5                 |


Если же объем прозводства всегда был постоянным, то получили бы следующие результаты

In [35]:
resPollutioni = 0
pollutioni = 0
resultProfit=0
for i in range(5):
    dirtyProfiti = B(6) 
    print('Прибыль до вычета штрафа за загрязнения в ', i+1, 'год', dirtyProfiti)
    
    pollutioni = W(6)
    print('Загрязнения в ', i+1, 'год', pollutioni)
        
    penaltyi = C(resPollutioni + pollutioni)
    print('Штраф за загрязнения на ', i+1, 'год', penaltyi)
    
    resPollutioni = A(resPollutioni + pollutioni)
    print('Остаточные загрязнения на ', i+2, 'год', resPollutioni)
    
    resultProfiti = dirtyProfiti - penaltyi
    print('Итоговая прибыль на ', i+1, 'год', resultProfiti, '\n')
    
    resultProfit +=resultProfiti
    
print('Итоговая прибыль по истечению 5 лет',resultProfit)

Прибыль до вычета штрафа за загрязнения в  1 год 10.189131601296467
Загрязнения в  1 год 8.77730198228669
Штраф за загрязнения на  1 год 1.298865751309214
Остаточные загрязнения на  2 год 5.553115992018953
Итоговая прибыль на  1 год 8.890265849987253 

Прибыль до вычета штрафа за загрязнения в  2 год 10.189131601296467
Загрязнения в  2 год 8.77730198228669
Штраф за загрязнения на  2 год 8.90492933394406
Остаточные загрязнения на  3 год 14.330417974305643
Итоговая прибыль на  2 год 1.2842022673524074 

Прибыль до вычета штрафа за загрязнения в  3 год 10.189131601296467
Загрязнения в  3 год 8.77730198228669
Штраф за загрязнения на  3 год 163.8422124059588
Остаточные загрязнения на  4 год 23.107719956592334
Итоговая прибыль на  3 год -153.65308080466235 

Прибыль до вычета штрафа за загрязнения в  4 год 10.189131601296467
Загрязнения в  4 год 8.77730198228669
Штраф за загрязнения на  4 год 2969.883750455242
Остаточные загрязнения на  5 год 31.885021938879024
Итоговая прибыль на  4 год -29

**Заметим, что при таком плане загрязнения дошли до отметки более, чем 40, а итоговая прибыль составила -56882.**